docs: support kmesh tls cert manager (#379)
kmesh-net/kmesh: docs/proposal/tls_cert.md

> ## Certificate Application and Management Module

Reviewer: ```suggestion
## Workload Certificate Management
```

Author: done

### Summary

> kmesh supports TLS capabilities that require the use of certificates issued by istiod, so a set of certificate application and management modules is needed to apply for certificates from istiod and manage the lifecycle of the certificates.

Reviewer: How do you translate these sensences,  gpt? Look not that native

Author: yes, and I checked again

### Motivation

> kmesh needs to provide TLS capabilities for the managed workloads, and it needs to be able to conveniently apply for and manage certificates, adding, deleting, and refreshing certificates at the appropriate time.

Author: OK

2. When kmesh restarts abnormally, all old certificate records are discarded, and all certificates are reapplied, without considering retaining the previous certificates.

### Proposal

Implement a certificate application module and a certificate management module, where:

Certificate Application Module:
- Establish an encrypted gRPC connection with istiod
- Construct a CSR request and the corresponding private key for the service account (sa) of the managed workload
- Use the CSR request to interact with istiod, and have istiod sign the certificate and return it

Certificate Management Module:
- Manage the timing of certificate operations: 1) When a new workload is added, 2) When a workload is deleted, 3) When a certificate is about to expire and needs to be automatically renewed
- Manage the storage and management of certificates
- Based on the certificate's expiration time, trigger the corresponding certificate refresh task when it's about to expire

### Constraints

> Currently, if you want to use kmesh's TLS capabilities, you need to modify the deployment when starting istio, and add `kmesh-system/kmesh` to the CA_TRUSTED_NODE_ACCOUNTS environment variable.

Reviewer 2: ```suggestion
Currently, if you want to use kmesh's TLS capabilities, you need to modify the deployment when starting istio, and add `kmesh-system/kmesh` to the `CA_TRUSTED_NODE_ACCOUNTS` environment variable.
```

Whether to start istio or istiod?

Author: done

## Design Details

### Certificate Application Module

Create a `caclient` client when kmesh starts, and establish an encrypted gRPC connection with istiod.

> Use the information in the workload to construct a CSR request and a private key, send the CSR request to istiod via the `caclient`, and have istiod sign and return the certificate.

Reviewer 2: what is `and have istiod sign` means? 
A long process can be illustrated in several steps

### Certificate Lifecycle Management

Use a channel, priority queue, and map to record and manage the certificates.

**Channel**: Manage certificate events, process certificate tasks based on Operation, create tasks in sequence from the channel to prevent some concurrency scheduling issues.

```go
chan: Used to receive all certificate-related events
type certRequest struct {
    Identity  string
    Operation int
}
```

Triggering Scenarios:
- When a new workload is added
- When a workload is deleted
- When a certificate is about to expire, take the certificate refresh task from the queue

> **Priority Queue**: Sort by certificate expiration time, keep the certificate that will expire soonest at the highest priority, periodically check the nearest expiring certificate and refresh it 10 minutes in advance.

Reviewer 3: 10 minutes -> 1 hour

Author: done

> **Priority Queue**: Sort by certificate expiration time, keep the certificate that will expire soonest at the highest priority, periodically check the nearest expiring certificate and refresh it 10 minutes in advance.

Reviewer 3: As discussed, maybe no need priority queue.

Author: add remaining issue

```go
Priority Queue Element:
type certExp struct {
    hostName string // Certificate name constructed using sa
    exp time.Time // Certificate expiration time
}
```

Update Scenarios:
- When adding a new certificate: Insert a new record
- When refreshing a certificate: Remove the old record, add a new record
- When deleting a certificate: Iterate and remove the old certificate record

**Map**: Record certificate information and certificate status:

```go
Map: Record the number of pods using the certificate
    key: Identity // Certificate name constructed using sa
    value: certCache

type certCache struct {
    cert istiosecurity.SecretItem // Certificate information
    refcnt uint32 // Record the number of pods using the certificate
}
```

> Update Scenarios:

Reviewer 3: Added the solution restrictions.

Author: add special design and remaining issue

- When the first pod managed by kmesh is added under a certain sa, a new certificate is added; a new record is created and added
- When all pods managed by kmesh under a certain sa are deleted (refCnt=0), the certificate is deleted; a record is deleted
- When a certificate is automatically refreshed upon expiration, update the value content; refresh the cert in the existing record

> - When a pod managed by kmesh is added under a certain sa, the corresponding refcnt is incremented, and refIp is recorded

Reviewer: refIp ? remove

Author: done

- When a pod managed by kmesh is deleted, the corresponding refcnt is decremented, and refIp is deleted

Lifecycle: The lifetime of the certificate for the entire sa; created when the sa certificate is applied for, deleted when the sa certificate is deleted

### Scenario 1: Adding a Certificate

> ![tls_cert_scenario1](pics/tls_cert_scenario1.svg)

Reviewer 2: Images can be centred

1. kmesh manages pod1, subscribes to the new workload, SecretManager looks for the certificate for the corresponding sa: if it already exists, the count is incremented; if it does not exist, a certificate application is initiated.

2. Construct and send the CSR request for sa1.

3. istiod issues the certificate.

4. Store the certificate:
   - Store the certificate
   - In the status information:
     - Record the count, count for this sa, record the number of pods using the certificate
   - Add a record with the expiration time to the priority queue

### Scenario 2: Deleting a Certificate

![tls_cert_scenario2](pics/tls_cert_scenario2.svg)

1. Delete pod1, delete the corresponding workload.

2. The sa count is decremented.

>    If the sa count is now 0, delete the certificate:

Reviewer 2: ```suggestion
   If the number of sa is 0 now, delete the certificate:
```
